src/ts/controller/core.ts

```diff
@@ -0,0 +1,263 @@
+
+/**
+ * Calculate collision group.
+ * @param {Array<Schedule|ScheduleViewModel>} schedules List of viewmodels.
+ * @returns {Array<number[]>} Collision Group.
+ */
+export function getCollisionGroup(schedules: Schedule[] | ScheduleViewModel[]) {
+  const collisionGroups: CollisionGroup = [];
+  let previousScheduleList: Array<Schedule | ScheduleViewModel>;
+
+  if (!schedules.length) {
+    return collisionGroups;
+  }
+
+  collisionGroups[0] = [schedules[0].cid()];
+  schedules.slice(1).forEach((schedule: Schedule | ScheduleViewModel, index: number) => {
+    previousScheduleList = schedules.slice(0, index + 1).reverse();
+
+    // If overlapping previous schedules, find a Collision Group of overlapping schedules and add this schedules
+    const found = previousScheduleList.find((previous: Schedule | ScheduleViewModel) => {
+      return schedule.collidesWith(previous);
+    });
+
+    if (!found) {
+      // This schedule is a schedule that does not overlap with the previous schedule, so a new Collision Group is constructed.
+      collisionGroups.push([schedule.cid()]);
+    } else {
+      collisionGroups
+        .slice()
+        .reverse()
+        .some(group => {
+          if (~inArray(found.cid(), group)) {
+            // If you find a previous schedule that overlaps, include it in the Collision Group to which it belongs.
+            group.push(schedule.cid());
+
+            return true; // returning true can stop this loop
+          }
+
+          return false;
+        });
+    }
+  });
+
+  return collisionGroups;
+}
+
+/**
+ * Get row length by column index in 2d matrix.
+ * @param {array[]} arr2d Matrix
+ * @param {number} col Column index.
+ * @returns {number} Last row number in column or -1
+ */
+export function getLastRowInColumn(arr2d: Array<any[]>, col: number) {
+  let row = arr2d.length;
+
+  while (row > 0) {
+    row -= 1;
+    if (!isUndefined(arr2d[row][col])) {
+      return row;
+    }
+  }
+
+  return -1;
+}
+
+/**
+ * Calculate matrix for appointment block element placing.
+ * @param {Collection} collection model collection.
+ * @param {Array<number[]>} collisionGroups Collision groups for schedule set.
+ * @returns {array} matrices
+ */
+export function getMatrices<T extends Schedule | ScheduleViewModel>(
+  collection: Collection<T>,
+  collisionGroups: CollisionGroup
+): ScheduleMatrix<T> {
+  const result: ScheduleMatrix<T> = [];
+
+  collisionGroups.forEach(function(group) {
+    const matrix: Array<T[]> = [[]];
+
+    group.forEach(function(scheduleID) {
+      const schedule: T = collection.items[scheduleID];
+      let col = 0;
+      let found = false;
+      let nextRow;
+      let lastRowInColumn;
+
+      while (!found) {
+        lastRowInColumn = getLastRowInColumn(matrix, col);
+
+        if (lastRowInColumn === -1) {
+          matrix[0].push(schedule);
+          found = true;
+        } else if (!schedule.collidesWith(matrix[lastRowInColumn][col])) {
```

Comment:
Similarly, check the parameter type of `collidesWith`.

Reply:
same as above I mentioned.